src/chainstate/stacks/boot/costs.clar

```diff
@@ -0,0 +1,495 @@
+    (runtime u1))
+
+(define-read-only (cost_analysis_option_cons (n uint))
+    (runtime u1))
+
+(define-read-only (cost_analysis_option_check (n uint))
+    (runtime u1))
+
+(define-read-only (cost_analysis_bind_name (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_list_items_check (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_check_tuple_get (n uint))
+    (runtime (logn n u1 u1)))
+
+(define-read-only (cost_analysis_check_tuple_cons (n uint))
+    (runtime (nlogn n u1 u1)))
+
+(define-read-only (cost_analysis_tuple_items_check (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_check_let (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_lookup_function (n uint))
+    (runtime u1))
+
+(define-read-only (cost_analysis_lookup_function_types (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_lookup_variable_const (n uint))
+    (runtime u1))
+
+(define-read-only (cost_analysis_lookup_variable_depth (n uint))
+    (runtime (nlogn n u1 u1)))
+
+(define-read-only (cost_ast_parse (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_ast_cycle_detection (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_analysis_storage (n uint))
+    {
+        runtime: (linear n u1 u1),
+        write_length: (linear n u1 u1),
+        write_count: u1,
+        read_count: u1,
+        read_length: u1
+    })
+
+(define-read-only (cost_analysis_use_trait_entry (n uint))
+    {
+        runtime: (linear n u1 u1),
+        write_length: (linear n u1 u1),
+        write_count: u0,
+        read_count: u1,
+        read_length: (linear n u1 u1)
+    })
+
+(define-read-only (cost_analysis_get_function_entry (n uint))
+    {
+        runtime: (linear n u1 u1),
+        write_length: u0,
+        write_count: u0,
+        read_count: u1,
+        read_length: (linear n u1 u1)
+    })
+
+(define-read-only (cost_analysis_fetch_contract_entry (n uint))
+    {
+        runtime: (linear n u1 u1),
+        write_length: u0,
+        write_count: u0,
+        read_count: u1,
+        read_length: (linear n u1 u1)
+    })
+
+(define-read-only (cost_lookup_variable_depth (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_lookup_variable_size (n uint))
+    (runtime (linear n u1 u0)))
+
+(define-read-only (cost_lookup_function (n uint))
+    (runtime u1))
+
+(define-read-only (cost_bind_name (n uint))
+    (runtime u1))
+
+(define-read-only (cost_inner_type_check_cost (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_user_function_application (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_let (n uint))
+    (runtime (linear n u1 u1)))
+
+(define-read-only (cost_if (n uint))
+    (runtime u1))
+
+(define-read-only (cost_asserts (n uint))
+    (runtime u1))
+
+(define-read-only (cost_map (n uint))
```

Comment:
What's `_n_` here -- is it the length of the list being mapped?  If so, then wouldn't this be linear?  Same question for `filter` and `fold` below.

Reply:
Yes, that's the size of the list. However, the "linear" cost of the map is charged during the execution of the function. This cost corresponds only to the constant overhead imposed by invoking the function through `map` (e.g., the function lookup cost).